cfpb/cfgov-refresh
# Send new users password reset email

Currently new users in Wagtail [must](https://github.com/cfpb/cfgov-refresh/blob/master/cfgov/v1/signals.py#L35) reset their password after their account has been created through the admin. Previously they'd have to try to login first with a temporary password and then request a reset email. This PR automatically sends a password reset email to users created through the Wagtail admin (`/admin/users/add/`).

## Additions

- Custom `WAGTAIL_USER_CREATION_FORM` setting overrides default admin form to use `v1.auth_forms.UserCreationForm`, which triggers a user email send on save.
- Added new `v1.email` module (and relevant tests) to contain email-related methods.
- Added new Django admin action to send password reset emails to users.

## Testing

- Run `tox` to verify that all Python unit tests pass, including the new ones in this PR.
- To manually test on a local server:
 1. Open the Wagtail admin to [http://localhost:8000/admin/users/] (or wherever you're running your dev server) 
 1. Click "Add a User" (in the upper right)
 1. Choose a username, email (doesn't matter for local testing), first and last name, and password
 1. Select roles (if desired) and then click "Add User"
 1. Note debug server output in console includes text of password reset email, including link to URL at [http://localhost:8000] (or wherever you were running your dev server)
- To verify that email links properly use the default Wagtail Site:
 1. Open the Wagtail admin to Site configuration at [http://localhost:8000/admin/sites/] (or wherever you're running your dev server)
 1. Change the default site to some custom hostname (e.g. `testhost`) and set the port to 443
 1. Click "Save" to commit your changes (no need to restart the dev server)
 1. Follow the manual test steps described above.
 1. Note debug server output now includes link to URL at [https://testhost](https://testhost) (or whatever hostname you chose)
 1. Revert default Site changes (if you wish) to restore previous functionality
- To verify the Django admin functionality:
 1. Open the Django admin to the user management page at [http://localhost:8000/django-admin/auth/user/] (or wherever you're running your dev server)
 1. Check the checkbox next to one or more users
 1. From the custom actions dropdown, choose "Send password reset email", and click "Go"
 1. Note debug server output showing password reset email(s)

## Review

- @cfpb/cfgov-backends 

## Screenshots

Sending password reset emails through the Django admin:

![image](https://cloud.githubusercontent.com/assets/654645/17306097/281e4650-57fc-11e6-8513-174738f6c984.png)

After sending the password reset emails:

![image](https://cloud.githubusercontent.com/assets/654645/17306104/35a24948-57fc-11e6-85e4-694959917eea.png)

## Notes

- Sending users email like this is complicated slightly by the need to know the hostname, port, and scheme of the server you want message links to go to. After investigating several alternatives, using the Wagtail [Site](http://docs.wagtail.io/en/v1.5.3/reference/pages/model_reference.html#site) framework seemed to be the most straightforward approach, especially because this framework is already being used elsewhere. This PR uses the `Site` with `is_default_site=True` (there can be only one) as the base URL to use for email links.

## Checklist
